# **Milestone** | College Football Data Analysis

<div style="text-align: center;">
<img src="https://upload.wikimedia.org/wikipedia/en/thumb/c/cf/NCAA_football_icon_logo.svg/2560px-NCAA_football_icon_logo.svg.png" alt="NCAA Football logo" width="200"/>
</div>


## Introduction
Welcome to the world of college football. For this Milestone, you're part of a data-driven sports analytics team, working with a major sports organization or university to help improve the performance of football teams. Your task is to dive into the data of Division I universities with football teams, extracting
insights that just might influence decisions on recruitment, conference performance, and enrollment strategies.

With the college football season in full swing, your work will help the organization answer key questions.

Let's begin by exploring the data at hand.

<div style="border: 3px solid #f8c43e; background-color: #fff3c1; padding: 15px; border-radius: 8px; color: #222; display: flex; align-items: center;">
  <span style="font-size: 10pt;">
      To import this data, remember to run the following cell.
  </span>
</div>



In [2]:
# load the data
import sys
sys.path.append('../files')

from ncaa_fbs import nickname, enrollment

## Data Exploration: Nicknames and Enrollments

We have two primary datasets: `nickname` and `enrollment`. The `nickname` dictionary provides the university name as keys and their respective team nicknames as values. The `enrollment` dictionary contains the university names as keys and their student enrollments as values.

Here are code snippets of each dicitonary to give you an understanding of their structure.

```python
nickname = {'Air Force': 'Falcons',
            'Akron': 'Zips',
            'Alabama': 'Crimson Tide',
            'Appalachian State': 'Mountaineers',
            'Arizona': 'Wildcats',
            'Arizona State': 'Sun Devils',
            'Arkansas': 'Razorbacks',
            'Arkansas State': 'Red Wolves',
            'Army': 'Black Knights',
            'Auburn': 'Tigers',
            'Ball State': 'Cardinals',
              ...
            'Wisconsin': 'Badgers',
            'Wyoming': 'Cowboys'}
```

The team from `Arizona State`, Dr. Crider's and Dr. Alvarez's alma mater, has the nickname `Sun Devils`.

```python
enrollment  = {'Air Force': 4181,
               'Akron': 14516,
               'Alabama': 38316,
               'Appalachian State': 20641,
               'Arizona': 49471,
               'Arizona State': 77881,
               'Arkansas': 29068,
               'Arkansas State': 12863,
               'Army': 4594,
               'Auburn': 31526,
               'Ball State': 19337,
                 ...
               'Wisconsin': 47932,
               'Wyoming': 11479}
```

The student population at `Arizona State` is `77,881` students.

To get an even better sense of this data, let's examine a few entries. Preview the data for `Purdue`, by passing this *key* to each one of the dictionaries and printing the associated *value*.

In [3]:
# Preview the data for Purdue
nick_name = nickname["Purdue"]
enroll_name = enrollment["Purdue"]
print(f"The nickname for Purdue is \"Boilermakers\" and they have a population of {enroll_name:,.0f} students.")

The nickname for Purdue is "Boilermakers" and they have a population of 49,639 students.


The nickname for Purdue is "Boilermakers" and have a student population of 49,639 students. Let's move on to
analyzing this information.

### Task 1: Team Count

As a first step, determine the total number of Division I Football Bowl
Subdivision (FBS) teams in our dataset. This will give us a sense of the
scope of our analysis. Store the result in a variable named `n_teams`.

Use an f-string to print out a complete sentence with the number of teams in the Division I FBS.

<div style="border: 3px solid #b67ae5; background-color: #f9f1ff; padding: 15px; border-radius: 8px; color: #222; display: flex; align-items: center;">
<span style="font-size: 10pt;">
<strong>Hint: </strong>Remember that the number of keys in any of the dictionaries will give you the total number of teams.
</span>
</div>





In [4]:
# How many FBS teams are there?
n_teams = len(nickname)

# print the required output
print(f"The number of schools in Division I FBS is {n_teams}.")

The number of schools in Division I FBS is 131.


### Task 2: Identifying Team Nicknames

Your team is interested in identifying all universities with the nickname "Tigers". Use a loop to iterate through each team in the `nickname` dictionary. For each team, use a conditional to check `if` their nickname is "Tigers". If so, append the university name to a Python list named `matches`.

Print `matches` to the screen.

<div style="border: 3px solid #b67ae5; background-color: #f9f1ff; padding: 15px; border-radius: 8px; color: #222; display: flex; align-items: center;">
<span style="font-size: 10pt;">
<strong>Hint: </strong>In order to access the school's nickname, you have to pass the name of the school as the key.
</span>
</div>



In [5]:
# set up an empty list for matching universities
matches = []

# loop over universities and append to list of matches when a match is found
for name in nickname:
    mascot = nickname[name]
    if mascot == "Tigers":
        matches.append(name)

# print the required output
print(f"The schools with Tigers as their nickname are: {matches}")

The schools with Tigers as their nickname are: ['Auburn', 'Clemson', 'LSU', 'Memphis', 'Missouri']


<div style="border: 3px solid #f8c43e; background-color: #fff3c1; padding: 15px; border-radius: 8px; color: #222; display: flex; align-items: center;">
  <span style="font-size: 10pt;">
      If done correctly, your output should be a list containing <strong>5 universities</strong>.
  </span>
</div>



### Task 3: University Enrollment

Next, you'll analyze the enrollment data to identify universities with enrollments smaller than 10,000 and those larger than 50,000 in order to determine if there are more FBS teams with a small student body or more FBS teams with a large student body.

Use a loop to iterate through each team in the `enrollment` dictionary.
*   If a university's enrollment is smaller than 10,000, increase the counting variable named `small_unis` by `1`.
*   If a university's enrollment is larger than 50,000, increase the counting variable named `large_unis` by 1.


In [6]:
# set up counter variables
small_unis = 0
large_unis = 0
medium_unis = 0

# loop over all universities
for student_count in enrollment:
    int_enroll = enrollment[student_count]
    if int_enroll < 10000:
        small_unis += 1
    elif int_enroll > 50000:
        large_unis += 1
    else:
        medium_unis += 1

# print the required outputs
print(f"There are {small_unis} schools with less than 10,000 students enrolled and {large_unis} schools with more than 50,000 students enrolled.")
print(f"There are also {medium_unis} schools that have more than 10,000 and less than or equal to 50,000 students enrolled.")

There are 7 schools with less than 10,000 students enrolled and 14 schools with more than 50,000 students enrolled.
There are also 110 schools that have more than 10,000 and less than or equal to 50,000 students enrolled.


Are there more FBS universities with enrollments smaller than 10,000, or with enrollments larger than 50,000?

*Yes, there are twice as many (14) universities with enrollments over 50k than there are universities with less than 10k students*

<div style="border: 3px solid #30EE99; background-color: #f0fff4; padding: 15px; border-radius: 8px; color: #222; display: flex; align-items: center;">
  <span style="font-size: 10pt;">
    <strong>Try This AI Prompt:</strong> How might university enrollment be relevant to a college football conference when making strategic decisions?
  </span>
</div>

In [7]:
# run this cell to import more data!
from ncaa_fbs import fbs_teams

### Task 4: Conference Enrollment Breakdown

The dictionary variable used in this task is called `fbs_teams`. In it, each university is now associated with a list of values, instead of one dictionary for each value.

For each university, the associated list contains items in this order:
* Index 0: Nickname
* Index 1: City
* Index 2: State
* Index 3: Enrollment
* Index 4: Conference

Here is a snippet of `fbs_teams` for a few entries in the dictionary:

```python
{ ...
 'Arizona State': ['Sun Devils', 'Tempe', 'Arizona', 77881, 'Pac-12'],
 'Arkansas': ['Razorbacks', 'Fayetteville', 'Arkansas', 29068, 'SEC'],
 'Arkansas State': ['Red Wolves', 'Jonesboro', 'Arkansas', 12863, 'Sun Belt'],
  ...
}
```

Just like before, preview the data for one university, `Arizona State`, by passing this *key* to the `fbs_teams` dictionary, and printing the associated *value*.

In [8]:
# preview the data for Arizona State
# print the whole list of values
print(fbs_teams["Arizona State"])

# print just the nickname
print(fbs_teams["Arizona State"][0])

# print just the city
print(fbs_teams["Arizona State"][1])

# print just the state
print(fbs_teams["Arizona State"][2])

# print just the enrollment
print(fbs_teams["Arizona State"][3])

# print just the conference
print(fbs_teams["Arizona State"][4])


['Sun Devils', 'Tempe', 'Arizona', 77881, 'Pac-12']
Sun Devils
Tempe
Arizona
77881
Pac-12


In the code above we can see that Arizona State are called the **Sun Devils**, are located in the city of **Tempe** in the state of **Arizona**. They have a total student population of **77,881**, and play in the **Pac-12** conference!

We'll use this dictionary to answer the following questions:
1. **Does the Big 12 conference actually have twelve schools?**  
2. **What is the total enrollment across these Big 12 universities?**
3. **What is the average enrollment across these Big 12 universities?**


One approach to answering these questions is to store the enrollment for each school in a `list`, called `students`, defined below.

Then the provided print statement will use different functions to calculate the values needed. Remember, each entry in `students` will correspond to a single university. It's `len`gth will be the number of universities in the Big 12. The `sum` will be the total enrolllment in the Big 12 conference. And the average enrollment can be calculated by combining the `len` and `sum` functions.

Use the code template below to complete Task 4.

In [9]:
# approach 1: using a list followed by summary functions
students = []
school_Big12 = []

# loop over universities, and append enrollments to the list
for school in fbs_teams:
    conference = fbs_teams[school][4]
    if conference == "Big 12":
        students.append(fbs_teams[school][3])

# print the statistics
print(f"1.\tThe Big 12 Conference has {len(students)} schools.")
print(f"2.\tThe total enrollment of the Big 12 schools is {sum(students):,.0f} students.")
print(f"2.\tThe average enrollment per school of the Big 12 schools is {sum(students) / len(students):,.0f} students.")

1.	The Big 12 Conference has 10 schools.
2.	The total enrollment of the Big 12 schools is 280,990 students.
2.	The average enrollment per school of the Big 12 schools is 28,099 students.


<div style="border: 3px solid #f8c43e; background-color: #fff3c1; padding: 15px; border-radius: 8px; color: #222; display: flex; align-items: center;">
  <span style="font-size: 10pt;">
      If done correctly, you will have **10** schools in the Big 12 conference. Weird, right?!?!
  </span>
</div>



A different approach is to use **accumulator** variables. An accumulator variable is a variable that is modified in each iteration of the loop. For example, if you are storing the number of students for each university, outside of the loop you will define `n_students = 0` but within the loop you will increment the value of `n_students` for every university that matches your criteria. Incrementing and assigning back to the same variable can be done with the `+=` operator.

In the code below, you have two accumulator variables, `n_students` and `n_universities` both set to zero intially. For each item in your loop that matches the criteria you will update the `n_students` variable based on the enrollment at that university and increment the `n_universities` by `1`.

The print statement provided at the end will show the results in the cell output.

In [53]:
# approach 2: using two numeric variables to accumulate the results
n_students = 0
n_universities = 0

# loop over universities, and accumulate values
for school in fbs_teams:
    conference = fbs_teams[school][4]
    if conference == "Big 12":
        n_students = n_students + fbs_teams[school][3]
        n_universities += 1

# print the results
print(f"1.\tThe Big 12 Conference has {n_universities} schools.")
print(f"2.\tThe total enrollment of the Big 12 schools is {n_students:,.0f} students.")
print(f"2.\tThe average enrollment per school of the Big 12 schools is {n_students / n_universities:,.0f} students.")

1.	The Big 12 Conference has 10 schools.
2.	The total enrollment of the Big 12 schools is 280,990 students.
2.	The average enrollment per school of the Big 12 schools is 28,099 students.


<div style="border: 3px solid #30EE99; background-color: #f0fff4; padding: 15px; border-radius: 8px; color: #222; display: flex; align-items: center;">
  <span style="font-size: 10pt;">
    <strong>Try This AI Prompt:</strong> I explored two distinct methods for aggregating data: using a list to store values and then calculating statistics versus using accumulator variables within a loop. Compare and contrast the two approaches. Explain a scenario where one approach might be preferred over the other.
  </span>
</div>

## Task 5: Highlighting Areas for Further Investigation

Based on your initial findings, what are 2-3 key questions or areas that the athletic department should explore further? What additional data would be valuable to gather?

Remember, you can use ChatGPT to brainstorm and structure your recommendations in a clear, professional manner!


Being that the athletic department is made up of athletes (quite redundant), data could be pulled showing the mean school major or degree plan of the top athletes. While we narrowed it down to the "Big 12" conference, the data is already there for multiple schools. It would be suggested to get a list of the top 25% of athletes (based on statistics of player performance), and find out what their major was. For example, if a player elects to have a degree plan of psychology, but a particular university doesn't even offer that degree plan, it would eliminate the school as a choice for the player.

The athletic department could also focus on funding for its athletic scholarships. Data like what funds are available, how to improve alumni participation and sponsorship, and the amount of private donations should be analyzed. This is because most of the top high school footbal players are 'recruited' or at least aware of what school they would like to attend. School attendance and choice is largely based on culture. So, for example, if a school in Kansas can make a very appealing scholarship offer to a student in Austin, TX, like a full-ride scholarship based on player performance, you may just influence the student to attend your school in Kansas instead of UT in Austin.

## LevelUp

Yet another way of storing the Division I universities information is by using a dictionary with *dictionaries* for values. In the task above you had a dictionary with a *list* for values. To get to the desired value, you would use the appropriate index.

In the case of a dictionary with dictionary values, you would instead use the *key* associated with the desired value.

A snippet of the `fbs_teams_dict` would look like:

```python
{...
 'Arizona': {'nickname': 'Wildcats',
             'city': 'Tucson',
             'state': 'Arizona',
             'enrollment': 49471,
             'conference': 'Pac-12'},
 'Arizona State': {'nickname': 'Sun Devils',
                   'city': 'Tempe',
                   'state': 'Arizona',
                   'enrollment': 77881,
                   'conference': 'Pac-12'},
 'Arkansas': {'nickname': 'Razorbacks',
              'city': 'Fayetteville',
              'state': 'Arkansas',
              'enrollment': 29068,
              'conference': 'SEC'},
...
```

<div style="border: 3px solid #f8c43e; background-color: #fff3c1; padding: 15px; border-radius: 8px; color: #222; display: flex; align-items: center;">
  <span style="font-size: 10pt;">
      To import this data, run the following cell.
  </span>
</div>

In [34]:
# import data
from ncaa_fbs import fbs_teams_dict

Compare this to the `fbs_teams` structure you used for Task 4. Store all the enrollment for Purdue University in a variable called `purdue_data`. Print `purdue_data` to the screen.


In [42]:
# Print out all the enrollment for Purdue University
purdue_data = fbs_teams_dict["Purdue"]
print(purdue_data)

{'nickname': 'Boilermakers', 'city': 'West Lafayette', 'state': 'Indiana', 'enrollment': 49639, 'conference': 'Big Ten'}


Notice that 'purdue_data' is itself a dictionary. If we want to get the enrollment for Purdue University, we still have 'Purdue' as the outer dictionary key, but now we specify the `'enrollment'` feature for the inner dictionary key:


In [49]:
# use the appropriate key to get the enrollment for Purdue University.
# recall that before we used the appropriate *index* to get the enrollment.
purdue_data = fbs_teams_dict["Purdue"]["enrollment"]
print(purdue_data)

49639


Your task is to replicate the conference enrollment analysis from Task 4, but using the `fbs_teams_dict` structure. This exercise will reinforce your understanding of nested dictionaries and data manipulation.

Specifically, calculate the following for the 'Big 12' conference:

1.  The number of universities.
2.  The total student enrollment.
3.  The average student enrollment.


<div style="border: 3px solid #b67ae5; background-color: #f9f1ff; padding: 15px; border-radius: 8px; color: #222; display: flex; align-items: center;">
<span style="font-size: 10pt;">
<strong>Note: </strong>The key order is really important when accessing nested dictionaries! Attempting to access the conference first will result in an error!!
</span>
</div>



In [57]:
# Task 4 -- approach 2: using two numeric variables to accumulate the statistics
n_students = 0
n_universities = 0

# loop over universities, and accumulate values
for school in fbs_teams_dict:
    conference_big12 = fbs_teams_dict[school]["conference"]
    if conference_big12 == "Big 12":
        n_students = n_students + fbs_teams_dict[school]["enrollment"]
        n_universities += 1

# print the statistics
print(f"1.\tThe Big 12 Conference has {n_universities} schools.")
print(f"2.\tThe total enrollment of the Big 12 schools is {n_students:,.0f} students.")
print(f"2.\tThe average enrollment per school of the Big 12 schools is {n_students / n_universities:,.0f} students.")

1.	The Big 12 Conference has 10 schools.
2.	The total enrollment of the Big 12 schools is 280,990 students.
2.	The average enrollment per school of the Big 12 schools is 28,099 students.
